### DISCUSSION TITLE
Wix Burn 3.11-> 4 rc4 Burn Bootstrapper Application how to use Engine.Apply

### DISCUSSION THREAD

**Discussion by woscostametaxa** [wixtoolset/issues#7300] [Questions] [answered]
Due to lack of any documentation so far I came across this problem:

Up until now I added a OnPlanComplete just like this:
```Cs
private void AfterPlanComplete(object sender, PlanCompleteEventArgs ea)
{
  if (ea.Status >= 0)
  {
    Engine.Apply(IntPtr.Zero);
  }
  else
  {
    ReturnValue = BootstrapperReturnValue.PlanError;
    Engine.Log(LogLevel.Error, $"Error during planning stage, error code {ea.Status}");
  }
}
```

which now gives me an error `Error 0x80070057: BA passed NULL hwndParent to Apply.`

Ok, now I can get the UI Window handle somehow . But I only have a window handle when I am started with Full UI or with /Passive - due to no documentation so far, after looking at externalengine.cpp:

```cpp
HRESULT ExternalEngineApply(
    __in BOOTSTRAPPER_ENGINE_CONTEXT* pEngineContext,
    __in_opt const HWND hwndParent
    )
{
    HRESULT hr = S_OK;
    BOOTSTRAPPER_ENGINE_ACTION* pAction = NULL;

    ExitOnNull(hwndParent, hr, E_INVALIDARG, "BA passed NULL hwndParent to Apply.");
    if (!::IsWindow(hwndParent))
    {
        ExitOnRootFailure(hr = E_INVALIDARG, "BA passed invalid hwndParent to Apply.");
    }

    pAction = (BOOTSTRAPPER_ENGINE_ACTION*)MemAlloc(sizeof(BOOTSTRAPPER_ENGINE_ACTION), TRUE);
    ExitOnNull(pAction, hr, E_OUTOFMEMORY, "Failed to alloc BOOTSTRAPPER_ENGINE_ACTION");

    pAction->dwMessage = WM_BURN_APPLY;
    pAction->apply.hwndParent = hwndParent;

    hr = EnqueueAction(pEngineContext, &pAction);
    ExitOnFailure(hr, "Failed to enqueue apply action.");

LExit:
    ReleaseMem(pAction);

    return hr;
}
```

what am I supposed to call Apply with in case of a silent installation?

All the old examples that I found for wix3 used the IntPtr.Zero all the time.

the TestBA.cs:
```CS
protected override void Run()
{
    this.dummyWindow = new Form();
    this.windowHandle = this.dummyWindow.Handle;

    this.Log("Running TestBA application");
    this.wait.Set();
    Application.Run();
}
```
 in the Wix4 just creates a new Form to get a dummy handle, is this how it is supposed to be handled in case of a silent installation?

**Comment by rseanhall** [comment, marked as answer]
The documentation is at https://wixtoolset.org/docs/fourthree/faqs/

**Reply by woscostametaxa** [reply]
Thank you, I presume that #5980 is the reason for that, and that the dummy handle should be sufficient for a first prototype when I never show any internal UIs.

**Reply by mistoll** [reply]
How is this supposed to work when using "-silent" on the bootstrapper?

**Reply by robmen** [reply]
Don't show the window.

**Reply by mistoll** [reply]
Thanks @robmen . That does the trick.
